Vizzuality/heco-invest
# feat: Integration with BC3 API

This one is a bit more complex and covers whole integration with BC3 API. Main parts which were developed as separate tasks and merged to this PR after they were approved are:
 - allow to switch between k.LAB API and old implementation which calculates impact values
 - submit context to k.LAB API when geometry of project changes
 - start polling tickets when context is submitted to k.LAB API
 - calculate impact when all demands from k.LAB API are obtained

Three new jobs which are responsible for handling this async communication are `Klab::SubmitContextsJob`, `Klab::PollImpactDemandsJob` and `Klab::CalculateImpactsJob`. All of them tries multiple times to send request to API endpoints, are able to recover from Faraday errors and wait between individual attempts. In case that response from BC3 is not obtained in defined time, error is reported to error notification system.

## Testing instructions

First of all, I have tried to cover all new features by unit tests so hopefully this manual testing does not need to be extra extensive. There are two use cases which should be verified:
When `KLAB_ENABLED` is false -- use old impact calculation:
 - find some project which already have calculated impact by old implementation. 
 - write down impact values somewhere and set all impact attributes of selected project to nil
 - trigger `recalculate_impacts` callback for selected project and verify that new impacts corresponds to old impacts which you have written down

When `KLAB_ENABLED` is true -- use BC3 integration
 - find some project and set its geometry to this geojson: `{"type":"Polygon","coordinates":[[[-72.67107290964314,3.5641500380320963],[-75.2281407807369,3.5641500380320963],[-75.2281407807369,5.302943221927137],[-72.67107290964314,5.302943221927137],[-72.67107290964314,3.5641500380320963]]]}`
 - verify that `Klab::SubmitContextsJob` is triggered
 - verify that `Klab::CalculateImpactsJob` is triggered
 - after some time (max couple minutes), attributes `#{impact_level}_demands_calculated` should be set to `true` and impacts should be calculated 
 - you can double check calculated impacts for appropriate impact level by running `Klab::Run.new.call` with geometry set to `context_string = Klab::BuildContextString.new(SELECTED_PROJECT, IMPACT_LEVEL).call`

## Tracking

https://vizzuality.atlassian.net/browse/LET-1317
